Upgrade typescript@3.5.3

markuskobler opened the pull request:
Fix's #78

markuskobler committed:
Upgrade typescript

asakusuma commented:
@krisselden can you review this?

krisselden commented on src/compiler/create-parse-config-host.ts:
```diff
@@ -11,6 +11,14 @@ export default function createParseConfigHost(
+    try {
+      return input.realpath(path) || path;
+    } catch {
+      return path;
```
need to throw if not the expected error `if (e.code !== "ENOENT") { throw e; }`

krisselden commented on src/compiler/create-parse-config-host.ts:
```diff
@@ -11,6 +11,14 @@ export default function createParseConfigHost(
 
+  function realpath(path: string): string  {
+    try {
+      return input.realpath(path) || path;
```
Is the `|| path` really needed in addition to the `try/catch`? There is no explanation of the failure scenario here.

krisselden reviewed (dismissed):
Need more explanation to how this is fixes #78 where @asakusuma mentioned that matchFiles changed, I expected some change related to that.

wycats commented:
Can we merge this so we can get up to date?

wycats commented:
It seems like TS thinks we're using private APIs? 

https://github.com/microsoft/TypeScript/pull/32564

krisselden commented:
We need to, it is too cumbersome to match the globbing behavior in the host interfaces without using that API, so that is not going to change.

krisselden commented:
I've asked them to make it public or easier.  It should be as easy to make a virtual fs like they use in their test suite with public api.

krisselden commented:
linking some related issues

https://github.com/microsoft/TypeScript/issues/13793
https://github.com/microsoft/TypeScript/issues/20878
https://github.com/microsoft/TypeScript/pull/21021

There is no reasonable way to deal with include/exclude globing in readDirectory in a host that is trying to gate all fs access without this API.

Broccoli treats input as immutable and actually builds inside of tmp directories, the only way to get node resolution and types to work is to pretend in the host like these tmp dirs are "mounted" in the project path.  In order to avoid using this API, I would have to parse the glob and match the under-specified glob behavior of matchFiles.

tomdale committed:
Don't catch input.realpath() exceptions

Since the underlying realpath implementation performs an existence check, there's no need to guard for ENOENT errors, and we can let exceptions propagate as they will represent a more serious error.

tomdale committed:
Upgrade to TypeScript v3.5.3
